pa11ycrawler/pipelines.py

```diff
@@ -20,89 +21,150 @@ class DuplicatesPipeline(object):
+        "Check for DRF urls."
+        url = URLObject(item["url"])
+        if url.path.startswith("/api/"):
+            raise DropItem("Dropping DRF url {url}".format(url=url))
+        else:
+            return item
 
-    def get_pa11y_settings(self, spider):
-        """
-        Sets `self.pa11y_settings` to the list of flags that should
-        be passed to the Pa11y CLI.
-        """
-        settings = []
-        for key, value in spider.settings.attributes.iteritems():
-            if is_pa11y_setting(key) and value.value:
-                flag_name = key.replace('_', '-').split('PA11Y-')[1].lower()
-                flag = '--{}="{}"'.format(flag_name, value.value)
-                settings.append(flag)
 
-        self.pa11y_settings = settings
-        return settings
+class Pa11yPipeline(object):
+    """
+    Runs the Pa11y CLI against `item['url']`, using the same request headers
+    used by Scrapy.
+    """
+    pa11y_path = "node_modules/.bin/pa11y"
+    cli_flags = {
+        "reporter": "1.0-json",
+    }
 
-    def _gen_pa11y_config(self, headers):
+    def write_pa11y_config(self, item):
         """
-        Creates tempfile with pa11y configuration.
-        This includes any headers that need to be kept.
-        Returns the file path of the new file.
+        The only way that pa11y will see the same page that scrapy sees
+        is to make sure that pa11y requests the page with the same headers.
+        However, the only way to configure request headers with pa11y is to
+        write them into a config file.
+
+        This function will create a config file, write the config into it,
+        and return a reference to that file.
         """
-        config = json.dumps({
+        config = {
             "page": {
-                "headers": headers,
+                "headers": item["request_headers"],
             },
-        })
-
-        pa11y_config_fp = tempfile.mkstemp(suffix='.json')[1]
-        with open(pa11y_config_fp, 'w') as pa11y_config_file:
-            pa11y_config_file.write(config)
-
-        return pa11y_config_fp
+        }
+        config_file = tempfile.NamedTemporaryFile(
+            mode="w",
+            prefix="pa11y-config-",
+            suffix=".json",
+            delete=False
+        )
+        json.dump(config, config_file)
+        config_file.close()
+        return config_file
+
+    def write_pa11y_results(self, item, results, data_dir):
+        """
+        Write the output from pa11y into a data file.
+        """
+        # `pa11y` outputs JSON, so we'll just add a bit more info for context
+        data = json.loads(results)
+        data.update(item)
+
+        # it would be nice to use the URL as the filename,
+        # but that gets complicated (long URLs, special characters, etc)
+        # so we'll make the filename a hash of the URL instead,
+        # and throw in the access time so that we can store the same URL
+        # multiple times in this data directory
```

Comment:
No, I just mis-read your description. When I heard 'hash' I was thinking md5 hash only. But reading the code more closely I see you're adding to that as a starting point. That's fine by me.